### PR TITLE
Refactored Synchroniser to use simpler self._futures construct

### PR DESCRIPTION
**Pull request by TarasLevelUp**
Will ease work on #34, as it remembers the exception, the synchroniser was killed with to raise them later.

### CODE REVIEW

File: src/asynqp/routing.py [hunk 1]
```diff
@@ -56,54 +57,48 @@ def handle(self, frame):
...
+            fut, bound_methods = self._futures[method].popleft()
+        except IndexError:
+            # XXX: we can't just ignore this.
+            log.error("Got an unexpected method notification %s", method)
+        else:
+            # Cleanup futures, that were awaited together, like
+            # (spec.BasicGetOK, spec.BasicGetEmpty)
+            for other_method in bound_methods:
+                if other_method != method:
+                    self._futures[other_method].remove((fut, bound_methods))
```
Comment (benjamin-hodgson) [inline src/asynqp/routing.py]: This is expensive. `deque.remove` has to search through the deque for the item it's looking for, and removing it means copying a potentially large number of other items into place. In other words `remove` is O(n), whereas it was amortised O(1) when it was based on `OrderedDict` before

Comment (Drizzt1991) [inline src/asynqp/routing.py]: Ok, I think I know how to fix this right away.

**Comment by Drizzt1991** [comment]
Fixed

**Comment by benjamin-hodgson** [comment]
GH won't let me merge it because there are conflicts. Please could you resolve them?

**Comment by TarasLevelUp** [comment]
done

**Comment by TarasLevelUp** [comment]
bump, if no more problems, merge this one pls.

**Pull request merged**